src/hotspot/share/prims/jvmtiEnvBase.cpp

```diff
@@ -756,6 +756,16 @@ JvmtiEnvBase::get_thread_state(oop thread_oop, JavaThread* jt) {
   return state;
 }
 
+jint
+JvmtiEnvBase::get_thread_state(oop thread_oop, JavaThread* jt) {
+  jint state = get_thread_state_base(thread_oop, jt);
+
+  if (is_passive_carrier_thread(jt, thread_oop)) {
+    state |= (JVMTI_THREAD_STATE_WAITING | JVMTI_THREAD_STATE_WAITING_INDEFINITELY);
```

Comment:
This is testing if the jt is carrying thread_oop and it's okay for the JVMTI state to reported as WAITING when waiting for something other than Object.wait.

One thing that is a bit confusing is the function name "is_passive_carrier_thread". A platform thread is either a carrier or not. Maybe for a different PR but I think is_passive_carrier_thread should be renamed to avoid the use of the word "passive".

Reply:
The lines 763-764 are to correct the state exactly for passive carrier thread, a carrier thread which can't progress until the execution control has not been returned from a virtual thread executed on the top. It is never for a platform thread which is not a carrier thread. "Passive" is the best word I was able to find for this meaning. Do you have any other word in mind?

Reply:
> The lines 763-764 are to correct the state exactly for passive carrier thread, a carrier thread which can't progress until the execution control has not been returned from a virtual thread executed on the top. It is never for a platform thread which is not a carrier thread. "Passive" is the best word I was able to find for this meaning. Do you have any other word/suggestion in mind?

It's just a carrier. A platform thread becomes a carrier when a virtual thread is mounted, it ceases to be a carrier once the virtual thread is unmounted. The mental model is that the carrier is blocked so reporting its state as waiting indefinitely is correct. Maybe you don't want to rename it in this PR but renaming this function to something like is_carrying would convey that it's asking the question if a given JavaThread is carrying the given virtual thread oop.

Reply:
Okay, I see you point. Unfortunately, I've always referred the platform thread with an executed FJP schedular as a carrier thread. The term 'carrier' with this meaning is everywhere in the JVMTI code. It looks very confusing to call a thread to be a carrier thread only during some phases of its execution.

Reply:
> Okay, I see you point. Unfortunately, I've always referred the platform thread with an executed FJP schedular as a carrier thread. The term 'carrier' with this meaning is everywhere in the JVMTI code. It looks very confusing to call a thread to be a carrier thread only during some phases of its execution.

Okay, I'm just pointing out that is_passive_carrier_thread is confusing looks a bit strange here as the is testing if a JavaThread is carrying a virtual thread oop -  it's not testing if the thread is owned by the virtual thread scheduler.

Reply:
I'm still thinking what identifier to use instead of `is_passive_carrier_thread`.
Just `is_carrier_thread` is going to be confusing as well.
What about `is_carrier_thread_waiting_for_virtual`?

Reply:
`is_carrying_carrier_thread`? a bit artificial, but it's a carrier thread and it's carrying a virtual thread

Reply:
> is_carrying_carrier_thread? a bit artificial, but it's a carrier thread and it's carrying a virtual thread

I guess, your suggestion is `is_carrying_virtual_thread`. Is it right?
If so, I like this suggestion.

Reply:
> > is_carrying_carrier_thread? a bit artificial, but it's a carrier thread and it's carrying a virtual thread
> 
> I guess, your suggestion is `is_carrying_virtual_thread`. Is it right? If so, I like this suggestion.

Up to you. I think any of this names is better than is_passive_carrier_thread.

Reply:
> I guess, your suggestion is `is_carrying_virtual_thread`. Is it right? If so, I like this suggestion.

Good, I think will be easy to understand at the use sites.

Reply:
Thank you, Alan and Alex.